### PR TITLE
[feat] refactor the interestform component

### PR DESCRIPTION
**Pull request by ccatherinetan**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
* refactor interest form to take in `listingData`, of type `Listing`, rather than `caseData` of type `CaseListing`
* created test page to see how diff listing type interests are displayed: http://localhost:3000/interest-test 
* refactored DateInput and RadioGroup to take in an optional prop `required` that adds a red asterisk to the label to indicate that a question is required. 

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @kyrenetam at the bottom."
Error
Screen Shot 2024-04-14 at 4 10 45 PM

Case 
Screen Shot 2024-04-14 at 4 10 31 PM

Case Interpretation / 1-time Interpretation
Screen Shot 2024-04-14 at 5 02 20 PM

Document Translation / LCA 
Screen Shot 2024-04-14 at 4 59 50 PM
LCA Error: Screen Shot 2024-04-14 at 5 00 17 PM

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
* check `components/InterestForm`
* styling: use the test page to see how diff listing type interests are displayed: http://localhost:3000/interest-test and `app/interest-test`
* check that the cases interest submits correctly
   * go to cases page (http://localhost:3000/cases) and submit an interest (make sure you're logged into an account with a profile created)
   * check supabase `interests` table to see if your interest has been added to the table 
* check that interests of all types submit correctly 
   * submit interests on http://localhost:3000/interest-test (change the listingData prop on `app/interest-test` for diff listing types)
   * check supabase to ensure that the listing type + responses match what you submitted 
   * note: if you use the dummy data on the the `interest-test` file, it will upsert the pre-existing interests submissions with those uuid's, so don't be alarmed if there isn't another row in the table. 

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
* potentially implement the interest form using Zod, a la @jinkang-0's onboarding pr: https://github.com/calblueprint/immigration-justice-project/pull/73 
    * currently, error handling is wacky and you can't submit interests for non-case listings. need to look into this
* confirm styling of "Your interest information is not saved! Please submit before leaving this page."
* test if it's possible to resubmit interest for the same case multiple times 

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @varortz, @kyrenetam

### CODE REVIEW

File: src/components/InterestForm/index.tsx [hunk 1]
```diff
@@ -86,46 +125,51 @@ export default function InterestForm({ caseData }: { caseData: CaseListing }) {
...
+            />
+          )}
+          {listingData.listing_type === 'CASE' && !interpretation && (
+            <RadioGroup
+              name="reason"
+              required
+              value={needsInterpreter}
+              setValue={setNeedsInterpreter}
+              options={['Yes', 'No']}
+              label="Do you need language interpretation help for the client?"
```
Comment (pragyakallanagoudar) [inline src/components/InterestForm/index.tsx]: Design question, but what if we rephrased this question to something like:

"Will you require language interpretation assistance to represent this client?"

(cc @kyrenetam @varortz)

File: src/components/InterestForm/index.tsx [hunk 2]
```diff
@@ -86,46 +125,51 @@ export default function InterestForm({ caseData }: { caseData: CaseListing }) {
...
+              error={
+                missingInfo
+                  ? 'Must select whether you need language support'
```
Comment (pragyakallanagoudar) [inline src/components/InterestForm/index.tsx]: Can we rephrase this error message so it uses the same wording as the question? i.e. "language interpretation help" or whatever we end up using in the question instead of "language support"

File: src/components/InterestForm/index.tsx [hunk 3]
```diff
@@ -2,72 +2,111 @@
...
+          }
+        }
+        responses = { ...responses, interest_reason: reason };
         const newInterest: Interest = {
-          listing_id: caseData.id,
-          listing_type: 'CASE',
+          listing_id: listingData.id,
+          listing_type:
+            listingData.listing_type === 'CASE' && interpretation
+              ? 'CASE_INT'
```
Comment (pragyakallanagoudar) [inline src/components/InterestForm/index.tsx]: It seems like even when I submit an interest just for a regular case (as an attorney), the listing type of the interest submission is `CASE_INT`? Is this expected behavior? i.e. does `CASE_INT` include both case interest and case interpretation interest?

File: src/components/InterestForm/index.tsx (hunk 1 above)
Comment (pragyakallanagoudar) [inline src/components/InterestForm/index.tsx]: Went into Supabase and manually changed some of the cases' `NEEDS_INTERPRETER` field to TRUE, but this question doesn't seem to show up. Curious if this is expected? What situation do we want this question (i.e. "Do you need language interpretation help for the client?") to show up in?

**Review by pragyakallanagoudar — COMMENTED** [review]
Looks really great, Catherine! I added a couple of questions on parts I was confused about, but largely the form looks fantastic and I love the new styling changes (especially the red asterisk!) :D

File: src/components/InterestForm/index.tsx (hunk 1 above)
Comment (varortz) [reply src/components/InterestForm/index.tsx]: yeah let's go ahead and update it

File: src/components/InterestForm/index.tsx (hunk 2 above)
Comment (varortz) [reply src/components/InterestForm/index.tsx]: let's update this to "Must select whether you need language interpretation assistance" to match Pragya's suggested wording for the question.

File: src/components/InterestForm/index.tsx [hunk 4]
```diff
@@ -86,46 +125,51 @@ export default function InterestForm({ caseData }: { caseData: CaseListing }) {
...
+              error={
+                missingInfo
+                  ? 'Must select whether you need language support'
+                  : ''
+              }
+            />
+          )}
+          <Flex $gap="10px" $direction="column">
+            {formQuestion(
+              'Why are you interested in this case?',
```
Comment (varortz) [inline src/components/InterestForm/index.tsx]: Could we rephrase the wording of the form label to "Why are you interested in this assignment?" when the listing is LCA or document translation?

File: src/components/InterestForm/index.tsx [hunk 5]
```diff
@@ -86,46 +125,51 @@ export default function InterestForm({ caseData }: { caseData: CaseListing }) {
...
-          </FormQuestion>
-          <FormTextArea
-            id="reason"
-            placeholder="I want to work on this case because..."
-            value={reason}
-            onChange={event => setReason(event.target.value)}
-          />
+            <FormTextArea
+              id="reason"
+              placeholder="I want to work on this case because..."
```
Comment (varortz) [inline src/components/InterestForm/index.tsx]: Similarly, could we rephrase the wording of the placeholder to "I want to work on this assignment because..." when the listing is LCA or document translation?

File: src/components/ListingDetails/index.tsx [hunk 6]
```diff
@@ -245,7 +248,7 @@ export default function ListingDetails({
   const interestSection = useMemo(() => {
     if (auth && auth.userId) {
       return profile?.profileData ? (
-        <InterestForm caseData={listingData as CaseListing} />
+        <InterestForm listingData={listingData} interpretation />
```
Comment (varortz) [inline src/components/ListingDetails/index.tsx]: can we update the interpretation prop from `interpretation` to `interpretation={interpretation}`. If we pass in `interpretation` alone, I think it's implicitly considered a truthy value and we run into the issue @pragyakallanagoudar mentioned in the comment above (for the cases page, I think we're getting the interest form meant for case interpretations).

File: src/components/InterestForm/index.tsx (hunk 1 above)
Comment (varortz) [reply src/components/InterestForm/index.tsx]: That question should be present for the cases page. I left a comment on `ListingDetails/index.tsx` that should address this!

File: src/components/InterestForm/index.tsx (hunk 3 above)
Comment (varortz) [reply src/components/InterestForm/index.tsx]: For a case listing (as an attorney) the listing type enum should be `CASE`. I left a comment on `ListingDetails/index.tsx` that should address this issue.

File: src/components/InterestForm/index.tsx [hunk 7]
```diff
@@ -2,72 +2,111 @@
...
+import RadioGroup from '@/components/RadioGroup';
 import DateInput from '../DateInput';
 import Button from '../Button';
 import {
   FormContainer,
   FormTextArea,
   FormQuestion,
-  RadioGroup,
-  Radio,
-  RadioInput,
```
Comment (varortz) [inline src/components/InterestForm/index.tsx:18]: Should we remove these styles from the `styles.tsx` file if we're using the `RadioGroup` component?

File: src/components/InterestForm/index.tsx [hunk 8]
```diff
@@ -86,46 +125,51 @@ export default function InterestForm({ caseData }: { caseData: CaseListing }) {
...
-          <FormQuestion $color={COLORS.greyDark}>
-            Why are you interested in this case?
-          </FormQuestion>
-          <FormTextArea
-            id="reason"
-            placeholder="I want to work on this case because..."
-            value={reason}
-            onChange={event => setReason(event.target.value)}
-          />
+            <FormTextArea
```
Comment (varortz) [inline src/components/InterestForm/index.tsx]: should we use the `TextAreaInput` component?

File: src/components/InterestForm/index.tsx (hunk 1 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx:149]: sg!

File: src/components/InterestForm/index.tsx (hunk 2 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: sg!

File: src/components/InterestForm/index.tsx (hunk 4 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: Do we want to keep 1-Time Interpretation as "case" or "assignment"? 

I was planning to change the line to: 

`Why are you interested in this ${listingData.listing_type === 'CASE' ? 'case': 'assignment'}?`,

The implementation above would have "assignment" for 1-time interpretations.

File: src/components/ListingDetails/index.tsx (hunk 6 above)
Comment (ccatherinetan) [reply src/components/ListingDetails/index.tsx]: AHHH gotcha!

File: src/components/InterestForm/index.tsx (hunk 1 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: Screen Shot 2024-04-17 at 2 00 54 PM
This is how it looks now. the question is a bit long though. until we have a better description, i'll keep it like this

File: src/components/InterestForm/index.tsx (hunk 5 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: sg! i have the same question about handling 1-time interpretation as the other comment

File: src/components/InterestForm/index.tsx (hunk 8 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: yes! changed!

File: src/components/InterestForm/index.tsx (hunk 1 above)
Comment (varortz) [reply src/components/InterestForm/index.tsx]: Yeah, I think it's ok if it looks a bit long, but let's check in with @kyrenetam  on this!

File: src/components/InterestForm/index.tsx (hunk 4 above)
Comment (varortz) [reply src/components/InterestForm/index.tsx]: Note from OH: Decided on "assignment", but we could either tailor it to One-Time Interpretation and refer to it as an "interpretation opportunity" or default to "assignment". Discuss at Thursday worksesh?

Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: consensus: default "opportunity"; cases → "case"

File: src/components/DateInput/index.tsx [hunk 9]
```diff
@@ -32,9 +36,12 @@ export default function DateInput({
...
   return (
     <InputDiv>
-      <InputLabel as="label" htmlFor={id}>
-        {label}
-      </InputLabel>
+      <Flex $direction="row">
+        <InputLabel as="label" htmlFor={id}>
+          {label}
+        </InputLabel>
+        {required && <InputLabel $color={COLORS.redMid}>&nbsp;*</InputLabel>}
```
Comment (jinkang-0) [inline src/components/DateInput/index.tsx]: instead of adding this, we could add the following style to `InputLabel`:
```css
&::after {
  content: ' *';
  color: ${COLORS.redMid};
}
```

File: src/components/InterestForm/index.tsx [hunk 10]
```diff
@@ -2,72 +2,94 @@
...
+    if (
+      (listingData.listing_type === 'CASE' &&
+        !interpretation &&
+        (startDate === '' || needsInterpreter === '')) ||
+      (((listingData.listing_type === 'CASE' && interpretation) ||
+        listingData.listing_type === 'INT') &&
+        startDate === '') ||
+      ((listingData.listing_type === 'LCA' ||
+        listingData.listing_type === 'DOC') &&
+        reason === '')
```
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: can we add some comments about the logic here?

Comment (jinkang-0) [reply src/components/InterestForm/index.tsx]: and for the if statements below

File: src/components/InterestForm/index.tsx [hunk 11]
```diff
@@ -2,72 +2,94 @@
...
+    ) {
       setMissingInfo(true);
       return;
     }
-    if (isValidDate(startDate)) {
+    let responses = {};
```
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: let's make this `const` and modify just the specific fields later
i.e. `responses.needs_interpreter = needsInterpreter === 'Yes'`

File: src/components/InterestForm/index.tsx [hunk 12]
```diff
@@ -86,45 +108,47 @@ export default function InterestForm({ caseData }: { caseData: CaseListing }) {
     <FormContainer>
       <H3>Submit Interest</H3>
       {submitted ? (
-        <P>We have received your submission!</P>
+        <P>Your submission has been received!</P>
```
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: can we add some padding bottom here? it looks a little squished atm
image

Comment (jinkang-0) [reply src/components/InterestForm/index.tsx]: `200px` should do

File: src/components/InterestForm/index.tsx (hunk 11 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: done!

File: src/components/InterestForm/index.tsx (hunk 10 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: yes! added! lmk if my comments are clarifying enough

File: src/components/InterestForm/index.tsx (hunk 12 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: sg, i changed it to `margin-bottom: auto` if that's ok? 
maybe it'll look better when the profile match is also there? 
EDIT: I saw your suggestion for 200px. i will implement that and push soon.

Comment (jinkang-0) [reply src/components/InterestForm/index.tsx]: actually that's a good point, we'll see and adjust after adding in profile match 👍
but having a `200px` minimum just in case won't hurt either

File: src/components/InterestForm/index.tsx [hunk 13]
```diff
@@ -1,73 +1,101 @@
...
+            listingData.listing_type === 'CASE' && interpretation
+              ? 'CASE_INT'
+              : listingData.listing_type,
           user_id: auth.userId,
           form_response: {
-            start_date: new Date(startDate),
-            interest_reason: reason,
-            // TODO: add needs_interpreter field
+            ...responses,
           },
```
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: instead of spreading `responses` as you've done here, we can just do `form_responses: responses`

Comment (jinkang-0) [reply src/components/InterestForm/index.tsx]: (mansplaining alert)
spreading technically creates a new object and copies the fields of `reponses` over, which is really only useful if we're adding extra fields besides those already in `responses`.
but since `responses` is all we're passing in, and we're not doing anything to `responses` after the fact, there's no need to use more compute resources to copy the object, we can just pass the pointer to the original `responses`

File: src/components/InterestForm/index.tsx [hunk 14]
```diff
@@ -1,73 +1,101 @@
...
-  RadioGroup,
-  Radio,
-  RadioInput,
-  FormFooter,
-  FormWarning,
-  ErrorText,
-  RadioLabel,
-} from './styles';
+import { useEffect, useState } from 'react';
+import { EmptySpace, FormContainer, FormFooter, FormWarning } from './styles';
```
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: instead of having these as individual imports, it may be good to do `import * as Styles from './styles';` just to make the markup clearer as to what components are local styles

then we can use it like `<Styles.EmptySpace> ... </Styles.EmptySpace>`

File: src/components/InterestForm/index.tsx [hunk 15]
```diff
@@ -1,73 +1,101 @@
...
+    ) {
       if (auth && auth.userId) {
+        if (
+          listingData.listing_type === 'CASE' ||
+          listingData.listing_type === 'INT'
+        ) {
+          responses.start_date = new Date(startDate);
+          if (listingData.listing_type === 'CASE' && !interpretation) {
+            responses.needs_interpreter = needsInterpreter === 'Yes';
+          }
```
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: let's add comments for these logic as well

File: src/components/InterestForm/index.tsx (hunk 10 above)
Comment (jinkang-0) [reply src/components/InterestForm/index.tsx]: yup, they're perfect

File: src/components/InterestForm/index.tsx (hunk 13 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: gotcha, done !

File: src/components/InterestForm/index.tsx (hunk 14 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: sg, done

File: src/styles/text.tsx [hunk 16]
```diff
@@ -34,12 +34,17 @@ export const CenteredH3 = styled(H3)`
...
   ${openSans.style}
   font-size: 1rem;
   font-weight: 600;
   color: ${props => props.$color || COLORS.greyDarker};
   margin: 0;
+  ${({ $required }) =>
+    $required &&
+    `&::after {
+  content: ' *';
+  color: ${COLORS.redMid}}`};
```
Comment (jinkang-0) [inline src/styles/text.tsx]: not sure if this makes sense for an `H4`, let's put the `required` prop in the respective `InputLabel` instead
or we can bring `InputLabel` into here since it is reused quite often

File: src/components/InterestForm/index.tsx (hunk 15 above)
Comment (ccatherinetan) [reply src/components/InterestForm/index.tsx]: added !

File: src/styles/text.tsx (hunk 16 above)
Comment (ccatherinetan) [reply src/styles/text.tsx]: mmm yeah i was also thinking that i was a little weird for H4, but i assumed that this was the main use of H4; good idea about inputlabel, though. 

done!

File: src/app/interest-test/page.tsx
Comment (jinkang-0) [inline src/app/interest-test/page.tsx:1]: let's remove this before we merge!

**Review by jinkang-0 — APPROVED** [review]
lgtm

File: src/components/DateInput/index.tsx [hunk 17]
```diff
@@ -32,7 +34,7 @@ export default function DateInput({
 
   return (
     <InputDiv>
-      <InputLabel as="label" htmlFor={id}>
+      <InputLabel $required={required} as="label" htmlFor={id}>
```
Comment (ccatherinetan) [inline src/components/DateInput/index.tsx]: use InputLabel from `styles/text` instead of `../styles`

**Review by varortz — APPROVED** [review]
Interest Form is looking great, thank you Catherine!
